browser/browser.c

```diff
@@ -1013,6 +1013,83 @@ static int file_tag(struct Menu *menu, int sel, int act)
   return ff->tagged - ot;
 }
 
+/*
+ * create_mailbox - Create a new mailbox
+ * @param state State used for handling IMAP case
+ * @param m_current The current mailbox to create the new mailbox under
+ * @retval 0 Successful creation
+ * @retval 1 Creation failed
+ * @retval 2 Creation aborted by user
+*/
+static int create_mailbox(struct BrowserState *state, struct Mailbox *m_current)
+{
+  int create_rc;
+
+#ifdef USE_IMAP /* leaving this until mbox_create implemented for IMAP */
+  if (state->imap_browse)
+  {
+    if (imap_mailbox_create(mutt_buffer_string(&LastDir)) == 0)
+    {
+      mutt_message(_("Mailbox created"));
+      return 0;
+    }
+    else
+    {
+      mutt_error(_("Mailbox creation failed"));
+      return 1;
+    }
+  }
+#endif /* USE_IMAP */
+
+  if (!m_current)
+    return 1;
+
+  if (m_current->mx_ops->mbox_create == NULL)
+  {
+    mutt_error(_("Create is not supported for %s mailboxes"), m_current->mx_ops->name);
+    return 1;
+  }
+
+  struct Buffer *path = mutt_buffer_pool_get();
+  if (mutt_buffer_get_field(_("Create mailbox: "), path, MUTT_COMP_NO_FLAGS,
+                            false, NULL, NULL, NULL) == -1)
+  {
+    mutt_buffer_pool_release(&path);
+    return 2;
+  }
+  struct Mailbox *new_mailbox = mutt_mem_malloc(sizeof(struct Mailbox));
+
+  create_rc = m_current->mx_ops->mbox_create(m_current->account, m_current,
+                                             mutt_buffer_string(path), &new_mailbox);
+
+  switch (create_rc)
+  {
+    case MX_CREATE_OK: /* = 0 (for imap_mailbox_create) */
+      mutt_message(_("Mailbox created"));
+      mutt_buffer_pool_release(&path);
+      return 0;
+    case MX_CREATE_EXISTS:
+      mutt_error(_("A %s mailbox already exists under %s"),
+                 m_current->mx_ops->name, mutt_buffer_string(path));
+      break;
+    case MX_CREATE_BAD_NAME:
+      mutt_error(_("Bad name '%s' (is it empty?)"), mutt_buffer_string(path));
+      break;
+    case MX_CREATE_SYS_ERROR:
+      mutt_error(_("Mailbox creation failed: %s"), strerror(errno));
+      break;
+    case MX_CREATE_ERROR:
+    case -1: /* imap_mailbox_create */
+      mutt_error(_("Mailbox creation failed"));
+      break;
+  }
+
+  /* Creation failed, cleanup */
+  mutt_buffer_pool_release(&path);
+  mutt_mem_free(&new_mailbox);
+  return 1;
```

Comment:
my C has become a bit rusty. Maybe this is cleaner with a flag to guard cleanup/put all the exits closer together:

```suggestion
  bool create_ok = false;
  switch (create_rc)
  {
    case MX_CREATE_OK: /* = 0 (for imap_mailbox_create) */
      mutt_message(_("Mailbox created"));
      create_ok = true
    case MX_CREATE_EXISTS:
      mutt_error(_("A %s mailbox already exists under %s"),
                 m_current->mx_ops->name, mutt_buffer_string(path));
      break;
    case MX_CREATE_BAD_NAME:
      mutt_error(_("Bad name '%s' (is it empty?)"), mutt_buffer_string(path));
      break;
    case MX_CREATE_SYS_ERROR:
      mutt_error(_("Mailbox creation failed: %s"), strerror(errno));
      break;
    case MX_CREATE_ERROR:
    case -1: /* imap_mailbox_create */
      mutt_error(_("Mailbox creation failed"));
      break;
  }
  mutt_buffer_pool_release(&path);

  if (create_ok) {
    return 0
  } else {
    mutt_buffer_pool_release(&path);
    mutt_mem_free(&new_mailbox);
    return 1;
  }
```